src/app/modules/user/user.service.ts

```diff
@@ -200,3 +238,198 @@ export const retrieveUser = (
     },
   )
 }
+
+// Private helper functions
+/**
+ * Retrieves the user with the given id.
+ * @param userId the id of the user to retrieve
+ * @returns ok(user document) if retrieval is successful
+ * @returns err(DatabaseError) if database errors occurs whilst retrieving user
+ * @returns err(MissingUserError) if user does not exist in the database
+ */
+const findAdminById = (
+  userId: string,
+): ResultAsync<IUserSchema, MissingUserError | DatabaseError> => {
+  return ResultAsync.fromPromise(UserModel.findById(userId).exec(), (error) => {
+    logger.error({
+      message: 'Database find user error',
+      meta: {
+        action: 'findAdminById',
+        userId,
+      },
+      error,
+    })
+    return new DatabaseError()
+  }).andThen((admin) => {
+    if (!admin) {
+      return errAsync(new MissingUserError())
+    }
+    return okAsync(admin)
+  })
+}
+
+/**
+ * Hashes both the given otp and contact data.
+ * @param otp the otp to hash
+ * @param contact the contact to hash
+ * @param logMeta additional metadata for logging, if available
+ * @returns ok(hashed otp and contact) is no errors occur
+ * @returns err(ApplicationError) if error occurs whilst hashing
+ *
+ */
+const hashOtpAndContact = (
+  otp: string,
+  contact: string,
+  logMeta?: Record<string, unknown>,
+): ResultAsync<
+  {
+    hashedOtp: string
+    hashedContact: string
+  },
+  ApplicationError
+> => {
+  return (
+    // Step 1: Hash OTP.
+    hashData(otp, logMeta)
+      // Step 2: Hash contact.
+      .andThen((hashedOtp) =>
+        hashData(contact, logMeta)
+          // If successful, return both hashedOtp and hashedContact.
+          .map((hashedContact) => ({
```

Comment:
does this `map` need to be nested? it's cleaner if it's outside right?

Reply:
Yes, since we need both `hashedOtp` and `hashedContact` in the same scope. If it's nested outside we will not have access to `hashedOtp`